chain/db/pubsub.go

```diff
@@ -2,95 +2,220 @@ package db
 
 import (
 	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"sort"
 	"strings"
 
+	"github.com/nknorg/nkn/chain/trie"
 	"github.com/nknorg/nkn/common/serialization"
 	"github.com/nknorg/nkn/transaction"
 	"github.com/nknorg/nkn/util/address"
 )
 
-func generateTopicKey(topic string) []byte {
-	topicKey := bytes.NewBuffer(nil)
-	topicKey.WriteByte(byte(PS_Topic))
-	serialization.WriteVarString(topicKey, strings.ToLower(topic))
-
-	return topicKey.Bytes()
+type pubSub struct {
+	subscriber string
+	meta       string
+	expiresAt  uint32
 }
 
-func generateTopicBucketKey(topic string, bucket uint32) []byte {
-	topicBucketKey := bytes.NewBuffer(nil)
-	topicKey := generateTopicKey(topic)
-	topicBucketKey.Write(topicKey)
-	serialization.WriteUint32(topicBucketKey, bucket)
-
-	return topicBucketKey.Bytes()
-}
+func (ps *pubSub) Serialize(w io.Writer) error {
+	if err := serialization.WriteVarString(w, ps.subscriber); err != nil {
+		return err
+	}
+	if err := serialization.WriteVarString(w, ps.meta); err != nil {
+		return err
+	}
 
-func generateSubscriberKey(subscriber []byte, identifier string, topic string, bucket uint32) []byte {
-	subscriberKey := bytes.NewBuffer(nil)
-	topicBucketKey := generateTopicBucketKey(topic, bucket)
-	subscriberKey.Write(topicBucketKey)
-	serialization.WriteVarBytes(subscriberKey, subscriber)
-	serialization.WriteVarString(subscriberKey, identifier)
+	if err := serialization.WriteUint32(w, ps.expiresAt); err != nil {
+		return err
+	}
 
-	return subscriberKey.Bytes()
+	return nil
 }
 
-func (cs *ChainStore) Subscribe(subscriber []byte, identifier string, topic string, bucket uint32, duration uint32, meta string, height uint32) error {
-	if duration == 0 {
-		return nil
-	}
+func (ps *pubSub) Deserialize(r io.Reader) error {
+	var err error
 
-	subscriberKey := generateSubscriberKey(subscriber, identifier, topic, bucket)
+	ps.subscriber, err = serialization.ReadVarString(r)
+	if err != nil {
+		return err
+	}
 
-	// PUT VALUE
-	err := cs.st.BatchPut(subscriberKey, []byte(meta))
+	ps.meta, err = serialization.ReadVarString(r)
 	if err != nil {
 		return err
 	}
 
-	err = cs.ExpireKeyAtBlock(height+duration, subscriberKey)
+	ps.expiresAt, err = serialization.ReadUint32(r)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (cs *ChainStore) Unsubscribe(subscriber []byte, identifier string, topic string, bucket uint32, duration uint32, height uint32) error {
-	if duration == 0 {
-		return nil
+func (ps *pubSub) Empty() bool {
+	return len(ps.subscriber) == 0 && len(ps.meta) == 0 && ps.expiresAt == 0
+}
+
+func getTopicId(topic string) []byte {
+	topicHash := sha256.Sum256([]byte(strings.ToLower(topic)))
+	return topicHash[:20]
```

Comment:
what about defining a constant instead of using hard coded 20 here?

Reply:
👍